src/connection_pool.rs

```diff
@@ -616,36 +764,41 @@ where
 fn rebalancer_loop<C>(max_connections: u32,
                       protected_data: ProtectedData<C>,
                       rebalance_check: RebalanceCheck,
-                      log: Logger)
+                      log: Logger,
+                      stop: Arc<AtomicBool>)
 where
     C: Connection
 {
-    loop {
+    let mut done = stop.load(AtomicOrdering::Relaxed);
+
+    while !done {
         let mut rebalance = rebalance_check.get_lock();
 
-        while !*rebalance {
-            rebalance = rebalance_check.condvar_wait(rebalance);
-        }
+        rebalance = rebalance_check.condvar_wait(rebalance);
 
-        // Briefly sleep in case the resolver notifies the pool of multiple
-        // changes within a short period
-        let sleep_time = time::Duration::from_millis(100);
```

Comment:
Same comment as below. You might consider adding a comment as to why we need to do this twice in the case that it is rebalancing.